File: Sources/CodexManager/Behavior/AccountIdentity.swift
import Foundation

enum AccountIdentity {
    private static let separator = "|"

    static func normalizedAccountID(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func normalizedEmail(_ value: String?) -> String? {
        guard let normalized = normalizedIdentifier(value) else { return nil }
        return normalized.contains("@") ? normalized.lowercased() : normalized
    }

    static func normalizedPrincipalID(
        _ principalID: String?,
        email: String?,
        accountID: String
    ) -> String {
        if let normalized = normalizedIdentifier(principalID) {
            return normalized.contains("@") ? normalized.lowercased() : normalized
        }
        if let normalizedEmail = normalizedEmail(email) {
            return normalizedEmail
        }
        return normalizedAccountID(accountID)
    }

    static func key(
        principalID: String?,
        email: String?,
        accountID: String
    ) -> String {
        "\(normalizedPrincipalID(principalID, email: email, accountID: accountID))\(separator)\(normalizedAccountID(accountID))"
    }

    static func key(for account: StoredAccount) -> String {
        key(principalID: account.principalID, email: account.email, accountID: account.accountID)
    }

    static func key(for account: AccountSummary) -> String {
        key(principalID: account.principalID, email: account.email, accountID: account.accountID)
    }

    static func key(for auth: ExtractedAuth) -> String {
        key(principalID: auth.principalID, email: auth.email, accountID: auth.accountID)
    }

    static func normalizedSelectionKey(_ value: String?) -> String? {
        normalizedIdentifier(value)
    }

    static func matches(_ stored: StoredAccount, extracted: ExtractedAuth) -> Bool {
        if key(for: stored) == key(for: extracted) {
            return true
        }

        guard normalizedAccountID(stored.accountID) == normalizedAccountID(extracted.accountID) else {
            return false
        }

        if let storedEmail = normalizedEmail(stored.email),
           let extractedEmail = normalizedEmail(extracted.email),
           storedEmail == extractedEmail {
            return true
        }

        return !hasExplicitPrincipal(principalID: stored.principalID, email: stored.email)
    }

    static func matches(_ lhs: StoredAccount, _ rhs: StoredAccount) -> Bool {
        if key(for: lhs) == key(for: rhs) {
            return true
        }

        guard normalizedAccountID(lhs.accountID) == normalizedAccountID(rhs.accountID) else {
            return false
        }

        if let lhsEmail = normalizedEmail(lhs.email),
           let rhsEmail = normalizedEmail(rhs.email),
           lhsEmail == rhsEmail {
            return true
        }

        return !hasExplicitPrincipal(principalID: lhs.principalID, email: lhs.email)
            || !hasExplicitPrincipal(principalID: rhs.principalID, email: rhs.email)
    }

    static func matches(selection: CurrentAccountSelection, account: StoredAccount) -> Bool {
        if let selectionKey = normalizedSelectionKey(selection.accountKey) {
            return selectionKey == key(for: account)
        }
        return normalizedAccountID(selection.accountID) == normalizedAccountID(account.accountID)
    }

    static func preferredMatchIndex(
        for extracted: ExtractedAuth,
        in accounts: [StoredAccount]
    ) -> Int? {
        let extractedKey = key(for: extracted)
        if let exactKeyIndex = accounts.firstIndex(where: { key(for: $0) == extractedKey }) {
            return exactKeyIndex
        }

        let normalizedExtractedAccountID = normalizedAccountID(extracted.accountID)
        let normalizedExtractedEmail = normalizedEmail(extracted.email)
        if let normalizedExtractedEmail,
           let emailMatchIndex = accounts.firstIndex(where: {
               normalizedAccountID($0.accountID) == normalizedExtractedAccountID
                   && normalizedEmail($0.email) == normalizedExtractedEmail
           }) {
            return emailMatchIndex
        }

        let wildcardMatches = accounts.indices.filter {
            normalizedAccountID(accounts[$0].accountID) == normalizedExtractedAccountID
                && !hasExplicitPrincipal(
                    principalID: accounts[$0].principalID,
                    email: accounts[$0].email
                )
        }
        guard wildcardMatches.count == 1 else {
            return nil
        }
        return wildcardMatches[0]
    }

    static func hasExplicitPrincipal(principalID: String?, email: String?) -> Bool {
        normalizedIdentifier(principalID) != nil || normalizedEmail(email) != nil
    }

    private static func normalizedIdentifier(_ value: String?) -> String? {
        guard let value else { return nil }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: apps/windows/src/shared/domain/account-identity.ts
import type { StoredAccount, AccountSummary, CurrentAccountSelection } from "../models/accounts";
import type { ExtractedAuth } from "../models/auth";

const separator = "|";

export interface AccountIdentityInput {
  principalId?: string;
  email?: string;
  accountId: string;
}

export function normalizedAccountId(value: string): string {
  return value.trim();
}

export function normalizedEmail(value: string | undefined): string | undefined {
  const normalized = normalizedIdentifier(value);
  if (!normalized) {
    return undefined;
  }
  return normalized.includes("@") ? normalized.toLowerCase() : normalized;
}

export function normalizedPrincipalId(
  principalId: string | undefined,
  email: string | undefined,
  accountId: string
): string {
  const normalized = normalizedIdentifier(principalId);
  if (normalized) {
    return normalized.includes("@") ? normalized.toLowerCase() : normalized;
  }

  const normalizedAccountEmail = normalizedEmail(email);
  if (normalizedAccountEmail) {
    return normalizedAccountEmail;
  }

  return normalizedAccountId(accountId);
}

export function accountKey(input: AccountIdentityInput): string {
  return `${normalizedPrincipalId(input.principalId, input.email, input.accountId)}${separator}${normalizedAccountId(
    input.accountId
  )}`;
}

export function accountKeyForStoredAccount(account: StoredAccount): string {
  return accountKey(account);
}

export function accountKeyForSummary(account: AccountSummary): string {
  return accountKey(account);
}

export function accountKeyForExtractedAuth(auth: ExtractedAuth): string {
  return accountKey(auth);
}

export function normalizedSelectionKey(value: string | undefined): string | undefined {
  return normalizedIdentifier(value);
}

export function storedAccountMatchesExtractedAuth(stored: StoredAccount, extracted: ExtractedAuth): boolean {
  if (accountKeyForStoredAccount(stored) === accountKeyForExtractedAuth(extracted)) {
    return true;
  }

  if (normalizedAccountId(stored.accountId) !== normalizedAccountId(extracted.accountId)) {
    return false;
  }

  const storedEmail = normalizedEmail(stored.email);
  const extractedEmail = normalizedEmail(extracted.email);
  if (storedEmail && extractedEmail && storedEmail === extractedEmail) {
    return true;
  }

  return !hasExplicitPrincipal(stored.principalId, stored.email);
}

export function storedAccountsMatch(lhs: StoredAccount, rhs: StoredAccount): boolean {
  if (accountKeyForStoredAccount(lhs) === accountKeyForStoredAccount(rhs)) {
    return true;
  }

  if (normalizedAccountId(lhs.accountId) !== normalizedAccountId(rhs.accountId)) {
    return false;
  }

  const lhsEmail = normalizedEmail(lhs.email);
  const rhsEmail = normalizedEmail(rhs.email);
  if (lhsEmail && rhsEmail && lhsEmail === rhsEmail) {
    return true;
  }

  return !hasExplicitPrincipal(lhs.principalId, lhs.email) || !hasExplicitPrincipal(rhs.principalId, rhs.email);
}

export function selectionMatchesAccount(selection: CurrentAccountSelection, account: StoredAccount): boolean {
  const selectionKey = normalizedSelectionKey(selection.accountKey);
  if (selectionKey) {
    return selectionKey === accountKeyForStoredAccount(account);
  }
  return normalizedAccountId(selection.accountId) === normalizedAccountId(account.accountId);
}

export function preferredMatchIndex(extracted: ExtractedAuth, accounts: readonly StoredAccount[]): number | undefined {
  const extractedKey = accountKeyForExtractedAuth(extracted);
  const exactKeyIndex = accounts.findIndex((account) => accountKeyForStoredAccount(account) === extractedKey);
  if (exactKeyIndex >= 0) {
    return exactKeyIndex;
  }

  const normalizedExtractedAccountId = normalizedAccountId(extracted.accountId);
  const normalizedExtractedEmail = normalizedEmail(extracted.email);
  if (normalizedExtractedEmail) {
    const emailMatchIndex = accounts.findIndex(
      (account) =>
        normalizedAccountId(account.accountId) === normalizedExtractedAccountId &&
        normalizedEmail(account.email) === normalizedExtractedEmail
    );
    if (emailMatchIndex >= 0) {
      return emailMatchIndex;
    }
  }

  const wildcardMatches = accounts
    .map((account, index) => ({ account, index }))
    .filter(
      ({ account }) =>
        normalizedAccountId(account.accountId) === normalizedExtractedAccountId &&
        !hasExplicitPrincipal(account.principalId, account.email)
    );

  return wildcardMatches.length === 1 ? wildcardMatches[0]?.index : undefined;
}

export function hasExplicitPrincipal(principalId: string | undefined, email: string | undefined): boolean {
  return normalizedIdentifier(principalId) !== undefined || normalizedEmail(email) !== undefined;
}

function normalizedIdentifier(value: string | undefined): string | undefined {
  const trimmed = value?.trim();
  return trimmed ? trimmed : undefined;
}
